File: frontend/src/services/stock.service.ts
import { apiService } from './api.service';
import type {
  StockResponse,
  StockMovementResponse,
  ReservationResponse,
  LowStockAlertResponse,
  StockAdjustmentRequest,
  ReserveStockRequest,
  CreateStockRequest,
  UpdateMinimumQuantityRequest,
  ApiResponse,
  PageResponse,
} from '../types/stock.types';

/**
 * Stock Service
 * Handles all stock-related API calls
 */
class StockService {
  private readonly BASE_PATH = '/stock-service/api';

  // ========== STOCK MANAGEMENT ==========

  /**
   * Get stock by product ID
   */
  async getStockByProductId(productId: number): Promise<StockResponse> {
    return apiService.get<StockResponse>(`${this.BASE_PATH}/stock/${productId}`);
  }

  /**
   * Get all stock with pagination
   */
  async getAllStock(page = 0, size = 20): Promise<PageResponse<StockResponse>> {
    return apiService.get<PageResponse<StockResponse>>(
      `${this.BASE_PATH}/stock?page=${page}&size=${size}`
    );
  }

  /**
   * Adjust stock quantity
   */
  async adjustStock(
    productId: number,
    request: StockAdjustmentRequest
  ): Promise<ApiResponse<StockResponse>> {
    return apiService.post<ApiResponse<StockResponse>>(
      `${this.BASE_PATH}/stock/${productId}/adjust`,
      request
    );
  }

  /**
   * Get stock movements for product
   */
  async getMovementsByProductId(
    productId: number,
    page = 0,
    size = 20
  ): Promise<PageResponse<StockMovementResponse>> {
    return apiService.get<PageResponse<StockMovementResponse>>(
      `${this.BASE_PATH}/stock/${productId}/movements?page=${page}&size=${size}`
    );
  }

  /**
   * Get low stock products
   */
  async getLowStockProducts(): Promise<StockResponse[]> {
    return apiService.get<StockResponse[]>(`${this.BASE_PATH}/stock/low-stock`);
  }

  // ========== RESERVATIONS ==========

  /**
   * Reserve stock for order
   */
  async reserveStock(request: ReserveStockRequest): Promise<ApiResponse<ReservationResponse>> {
    return apiService.post<ApiResponse<ReservationResponse>>(
      `${this.BASE_PATH}/stock/reserve`,
      request
    );
  }

  /**
   * Confirm reservation (order paid)
   */
  async confirmReservation(reservationId: number): Promise<ApiResponse<ReservationResponse>> {
    return apiService.post<ApiResponse<ReservationResponse>>(
      `${this.BASE_PATH}/stock/confirm/${reservationId}`
    );
  }

  /**
   * Release reservation (order cancelled)
   */
  async releaseReservation(reservationId: number): Promise<ApiResponse<ReservationResponse>> {
    return apiService.post<ApiResponse<ReservationResponse>>(
      `${this.BASE_PATH}/stock/release/${reservationId}`
    );
  }

  /**
   * Get reservations by order ID
   */
  async getReservationsByOrderId(orderId: number): Promise<ReservationResponse[]> {
    return apiService.get<ReservationResponse[]>(
      `${this.BASE_PATH}/stock/reservations/${orderId}`
    );
  }

  // ========== ADMIN ==========

  /**
   * Create stock for new product
   */
  async createStock(request: CreateStockRequest): Promise<ApiResponse<StockResponse>> {
    return apiService.post<ApiResponse<StockResponse>>(
      `${this.BASE_PATH}/admin/stock`,
      request
    );
  }

  /**
   * Update minimum quantity threshold
   */
  async updateMinimumQuantity(
    productId: number,
    request: UpdateMinimumQuantityRequest
  ): Promise<ApiResponse<StockResponse>> {
    return apiService.put<ApiResponse<StockResponse>>(
      `${this.BASE_PATH}/admin/stock/${productId}/minimum`,
      request
    );
  }

  /**
   * Get active low stock alerts
   */
  async getActiveLowStockAlerts(): Promise<LowStockAlertResponse[]> {
    return apiService.get<LowStockAlertResponse[]>(`${this.BASE_PATH}/admin/stock/alerts`);
  }
}

// Export singleton instance
export const stockService = new StockService();
export default stockService;

File: stock-service/src/main/java/com/retailsports/stock_service/service/StockService.java
package com.retailsports.stock_service.service;

import com.retailsports.stock_service.dto.request.CreateStockRequest;
import com.retailsports.stock_service.dto.request.ReserveStockRequest;
import com.retailsports.stock_service.dto.request.StockAdjustmentRequest;
import com.retailsports.stock_service.dto.request.UpdateMinimumQuantityRequest;
import com.retailsports.stock_service.dto.response.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface per la gestione dello stock
 */
public interface StockService {

    // ========== GESTIONE STOCK ==========
    
    /**
     * Ottieni stock per productId
     */
    StockResponse getStockByProductId(Long productId);
    
    /**
     * Ottieni tutti gli stock con paginazione
     */
    Page<StockResponse> getAllStock(Pageable pageable);
    
    /**
     * Aggiusta quantità stock (IN/OUT/ADJUSTMENT)
     */
    StockResponse adjustStock(Long productId, StockAdjustmentRequest request);
    
    /**
     * Ottieni storico movimenti per prodotto
     */
    Page<StockMovementResponse> getMovementsByProductId(Long productId, Pageable pageable);
    
    /**
     * Ottieni prodotti con scorte basse
     */
    List<StockResponse> getLowStockProducts();
    
    // ========== PRENOTAZIONI ==========
    
    /**
     * Prenota stock per ordine
     */
    ReservationResponse reserveStock(ReserveStockRequest request);
    
    /**
     * Conferma prenotazione (ordine pagato)
     */
    ReservationResponse confirmReservation(Long reservationId);
    
    /**
     * Rilascia prenotazione (ordine cancellato)
     */
    ReservationResponse releaseReservation(Long reservationId);
    
    /**
     * Ottieni prenotazioni per ordine
     */
    List<ReservationResponse> getReservationsByOrderId(Long orderId);
    
    // ========== ADMIN ==========
    
    /**
     * Crea stock per nuovo prodotto
     */
    StockResponse createStock(CreateStockRequest request);
    
    /**
     * Aggiorna soglia minima
     */
    StockResponse updateMinimumQuantity(Long productId, UpdateMinimumQuantityRequest request);
    
    /**
     * Ottieni alert scorte basse attivi
     */
    List<LowStockAlertResponse> getActiveLowStockAlerts();
}
